improve cross-compiling when bootstrapping the first glibc with a stage1 gcc

From 95f5a9a866695da4e038aa4e6ccbbfd5d9cf63b7 Mon Sep 17 00:00:00 2001
From: Joseph Myers <joseph@codesourcery.com>
Date: Tue, 3 Jul 2012 19:14:59 +0000
Subject: [PATCH] Avoid use of libgcc_s and libgcc_eh when building glibc.

---
 ports/ChangeLog.arm        |  7 +++++
 ports/sysdeps/arm/Makefile |  8 ++++++
 7 files changed, 206 insertions(+), 21 deletions(-)
 create mode 100644 elf/static-stubs.c

2012-07-03  Joseph Myers  <joseph@codesourcery.com>

	* Makeconfig [!+link] (+link-before-libc): New variable.
	[!+link] (+link-after-libc): Likewise.
	[!+link] (+link-tests): Likewise.
	[!+link] (+link): Define in terms of $(+link-before-libc) and
	$(+link-after-libc).
	[!+link-static] (+link-static-before-libc): New variable.
	[!+link-static] (+link-static-after-libc): Likewise.
	[!+link-static] (+link-static-tests): Likewise.
	[!+link-static] (+link-static): Define in terms of
	$(+link-static-before-libc) and $(+link-static-after-libc).
	[build-shared] (link-libc-before-gnulib): New variable.
	[build-shared] (link-libc-tests): Likewise.
	[build-shared] (link-libc): Define in terms of
	$(link-libc-before-gnulib).
	[!build-shared] (link-libc-tests): New variable.
	(link-libc-static-tests): New variable.
	[!gnulib] (gnulib-arch): New variable.
	[!gnulib] (gnulib-tests): Likewise.
	[!gnulib] (static-gnulib-arch): Likewise.
	[!gnulib] (static-gnulib-tests): Likewise.
	[!gnulib] (gnulib): Use $(gnulib-arch).  Do not use $(libgcc_eh).
	Define with "=" instead of ":=".
	[!gnulib] (static-gnulib): Use $(static-gnulib-arch).  Do not use
	-lgcc_eh $(libunwind).  Define with "=" instead of ":=".
	* Rules (binaries-all-notests): New variable.
	(binaries-all-tests): Likewise.
	(binaries-static-notests): Likewise.
	(binaries-static-tests): Likewise.
	(binaries-all): Define using $(binaries-all-notests) and
	$(binaries-all-tests).
	(binaries-static): Define using $(binaries-static-notests) and
	$(binaries-static-tests).
	(binaries-shared-tests): New variable.
	(binaries-shared-notests): Likewise.
	(binaries-shared): Remove variable.
	($(addprefix $(objpfx),$(binaries-shared-notests))): New rule.
	($(addprefix $(objpfx),$(binaries-shared-tests))): Likewise.
	($(addprefix $(objpfx),$(binaries-shared))): Remove rule.
	($(addprefix $(objpfx),$(binaries-static-notests))): New rule.
	($(addprefix $(objpfx),$(binaries-static-tests))): Likewise.
	($(addprefix $(objpfx),$(binaries-static))): Remove rule.
	* elf/Makefile (sln-modules): New variable.
	(extra-objs): Add $(sln-modules:=.o).
	(ldconfig-modules): Add static-stubs.
	($(objpfx)sln): Depend on $(sln-modules:%=$(objpfx)%.o).
	* elf/static-stubs.c: New file.

diff --git a/ports/ChangeLog.arm b/ports/ChangeLog.arm
index c411e1c..71e8b1b 100644
--- a/ports/ChangeLog.arm
+++ b/ports/ChangeLog.arm
@@ -1,3 +1,10 @@
+2012-07-03  Joseph Myers  <joseph@codesourcery.com>
+
+	* sysdeps/arm/Makefile (gnulib-arch): New variable.
+	(static-gnulib-arch): Likewise.
+	[subdir = elf] ($(objpfx)libgcc-stubs.a): New rule.
+	[subdir = elf] (lib-noranlib): Depend on $(objpfx)libgcc-stubs.a.
+
 2012-06-12  Joseph Myers  <joseph@codesourcery.com>
 
 	* sysdeps/unix/sysv/linux/arm/setcontext.S (__startcontext): Use
diff --git a/ports/sysdeps/arm/Makefile b/ports/sysdeps/arm/Makefile
index 5fd5136..3bdb6b3 100644
--- a/ports/sysdeps/arm/Makefile
+++ b/ports/sysdeps/arm/Makefile
@@ -1,8 +1,16 @@
+gnulib-arch = $(elfobjdir)/libgcc-stubs.a
+static-gnulib-arch = $(elfobjdir)/libgcc-stubs.a
+
 ifeq ($(subdir),elf)
 sysdep-dl-routines += tlsdesc dl-tlsdesc
 sysdep_routines += aeabi_unwind_cpp_pr1 find_exidx tlsdesc dl-tlsdesc
 sysdep-rtld-routines += aeabi_unwind_cpp_pr1 tlsdesc dl-tlsdesc
 shared-only-routines += aeabi_unwind_cpp_pr1
+
+$(objpfx)libgcc-stubs.a: $(objpfx)aeabi_unwind_cpp_pr1.os
+	$(build-extra-lib)
+
+lib-noranlib: $(objpfx)libgcc-stubs.a
 endif
 
 ifeq ($(subdir),csu)
-- 
1.7.12.4

